website/static/js/apiApplication.js

```diff
@@ -0,0 +1,317 @@
+/*
+ *  Knockout views for the Developer application pages: List and Create/Edit pages.
+ */
+
+
+'use strict';
+
+// TODO: Some of these may not be actually required. Dependency list ripped from profile view
+var $ = require('jquery');
+var ko = require('knockout');
+var bootbox = require('bootbox');
+require('knockout.validation');
+require('knockout.punches');
+ko.punches.enableAll();
+require('knockout-sortable');
+
+var $osf = require('./osfHelpers');
+var koHelpers = require('./koHelpers');
+var Raven = require('raven-js');
+require('js/objectCreateShim');
+
+
+/*
+ *  Store the data related to a single API application
+ */
+var ApplicationData = function (data){
+    var self = this;
+
+    self.clientSecret = ko.observable();
+
+    self.owner = ko.observable();
+    self.name = ko.observable().extend({required: true});
+    self.description = ko.observable();
+    self.homeUrl = ko.observable().extend({
+        url:true,
+        ensureHttp:true,
+        required:true
+        });
+    self.callbackUrl = ko.observable().extend({
+        url:true,
+        ensureHttp:true,
+        required:true
+    });
+
+    // Unchanging properties
+    self.clientId = null;
+    self.create_date = null;
+    self.detailUrl = null;
+    self.apiUrl = null;
+
+    // Load in data
+    if (data){
+        self.fromJSON(data);
+    }
+
+    // Enable value validation in form
+    var validated = ko.validatedObservable(self);
+    self.isValid = ko.computed(function(){
+        return validated.isValid();
+    });
+
+};
+
+// Serialize data for POST request
+ApplicationData.prototype.toJSON = function () {
+    var self = this;
+    return {
+        client_id: self.clientId,
+        client_secret: self.clientSecret(),
+        owner: self.owner(),
+        name: self.name(),
+        description: self.description(),
+        home_url: self.homeUrl(),
+        callback_url: self.callbackUrl()
+    };
+    };
+
+// Load data from JSON
+ApplicationData.prototype.fromJSON = function (data) {
```

Comment:
This might make sense as a class method that returns an `ApplicationData` instance--I think this would clarify the confusion here: https://github.com/CenterForOpenScience/osf.io/pull/3082/files#r34273201